internal/allocator/allocation.go

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier:Apache-2.0
+
+package allocator // import "go.universe.tf/metallb/internal/allocator"
+
+import (
+	"fmt"
+	"net"
+
+	"go.universe.tf/metallb/internal/ipfamily"
+	v1 "k8s.io/api/core/v1"
+)
+
+// Allocation represents a mapping of IP families (IPv4, IPv6, or both) to their respective IP addresses.
+// This type is used to select and organize IPs based on a service's IP family policy.
+// Depending on the family policy, the map may contain:
+// - Just an IPv4 address
+// - Just an IPv6 address
+// - Both IPv4 and IPv6 addresses for dual-stack support.
+//
+// Key:
+// - ipfamily.IPv4: Corresponds to the IPv4 address allocated for the service.
+// - ipfamily.IPv6: Corresponds to the IPv6 address allocated for the service.
+type Allocation struct {
+	PoolName string
+	IPV4     net.IP
+	IPV6     net.IP
+}
+
+func newAllocation(poolName string) *Allocation {
+	return &Allocation{
+		PoolName: poolName,
+		IPV4:     nil,
+		IPV6:     nil,
+	}
+}
+
+func (a *Allocation) getIPV4() net.IP {
+	ip := a.IPV4
+	if ip == nil {
+		return nil
+	}
+	if ip.To4() == nil {
```

Comment:
I have mixed feelings about this check. I'd say we should trust who filled this (still us) so in case of error let it explode.